ddht/subscription_manager.py

```diff
@@ -0,0 +1,71 @@
+import collections
+import logging
+from typing import Any, AsyncIterator, DefaultDict, NamedTuple, Optional, Set, Type
+
+from async_generator import asynccontextmanager
+from eth_typing import NodeID
+import trio
+
+from ddht.abc import SubscriptionManagerAPI
+from ddht.base_message import AnyInboundMessage, InboundMessage, TMessage
+from ddht.endpoint import Endpoint
+
+
+class _Subcription(NamedTuple):
+    send_channel: trio.abc.SendChannel[AnyInboundMessage]
+    filter_by_endpoint: Optional[Endpoint]
+    filter_by_node_id: Optional[NodeID]
+
+
+class SubscriptionManager(SubscriptionManagerAPI[TMessage]):
+    logger = logging.getLogger("ddht.SubscriptionManager")
+
+    _subscriptions: DefaultDict[Type[Any], Set[_Subcription]]
+
+    def __init__(self) -> None:
+        self._subscriptions = collections.defaultdict(set)
+
+    def feed_subscriptions(self, message: AnyInboundMessage) -> None:
+        message_type = type(message.message)
+
+        subscriptions = tuple(self._subscriptions[message_type])
+        self.logger.debug(
+            "Handling %d subscriptions for message: %s", len(subscriptions), message,
+        )
+        for subscription in subscriptions:
+            if subscription.filter_by_endpoint is not None:
+                if message.sender_endpoint != subscription.filter_by_endpoint:
+                    continue
+            if subscription.filter_by_node_id is not None:
+                if message.sender_node_id != subscription.filter_by_node_id:
+                    continue
+
+            try:
+                subscription.send_channel.send_nowait(message)  # type: ignore
+            except trio.WouldBlock:
+                self.logger.debug(
```

Comment:
Dropping these messages is a reasonable decision, but this behavior might deserve a warning message in a docstring!